File: baserecycleradapter/src/main/java/com/gigigo/baserecycleradapter/debouncedlisteners/DebouncedClickHandler.java
package com.gigigo.baserecycleradapter.debouncedlisteners;

import android.os.SystemClock;
import android.view.View;

public class DebouncedClickHandler {

    private final static long MINIMUM_INTERVAL_MILLIS = 300;

    private long intervalMillis = MINIMUM_INTERVAL_MILLIS;
    private long previousClickTimestamp;
    private final DebouncedListener debouncedOnClickListener;

    public DebouncedClickHandler(DebouncedListener debouncedOnClickListener) {
        this.debouncedOnClickListener = debouncedOnClickListener;
    }

    public DebouncedClickHandler(DebouncedListener debouncedOnClickListener, long intervalMillis) {
        this.debouncedOnClickListener = debouncedOnClickListener;
        this.intervalMillis = intervalMillis;
    }

    public boolean invokeDebouncedClick(int position, View view) {
        long currentTimestamp = SystemClock.uptimeMillis();
        boolean handled = false;
        if (currentTimestamp - previousClickTimestamp > intervalMillis) {
            handled = debouncedOnClickListener.onDebouncedClick(view, position);
            previousClickTimestamp = currentTimestamp;
        }
        return handled;
    }
}

File: baserecycleradapter/src/main/java/com/gigigo/baserecycleradapter/debouncedlisteners/DebouncedClickHandler.kt
package com.gigigo.baserecycleradapter.debouncedlisteners

import android.os.SystemClock
import android.view.View

open class DebouncedClickHandler(
    intervalMillis: Long? = MINIMUM_INTERVAL_MILLIS,
    private val debouncedOnClickListener: (Int, View) -> Boolean
) {

    var intervalMillis = MINIMUM_INTERVAL_MILLIS
    var previousClickTimestamp: Long = 0

    init {
        intervalMillis?.let { this.intervalMillis = it }
    }

    fun invoke(position: Int, view: View): Boolean {
        val currentTimestamp = SystemClock.uptimeMillis()
        var handled = false
        if (currentTimestamp - previousClickTimestamp > intervalMillis) {
            handled = debouncedOnClickListener.invoke(position, view)
            previousClickTimestamp = currentTimestamp
        }
        return handled
    }

    companion object {
        private const val MINIMUM_INTERVAL_MILLIS: Long = 300
    }
}
